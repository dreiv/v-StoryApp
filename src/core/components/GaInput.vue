<script setup lang="ts">
import { computed, useCssModule, useTemplateRef, type InputHTMLAttributes } from 'vue'

defineOptions({ inheritAttrs: false })

export interface GaInputProps extends /* @vue-ignore */ InputHTMLAttributes {
  invalid?: boolean
  success?: boolean
}
const { invalid, success } = defineProps<GaInputProps>()
const inputRef = useTemplateRef<HTMLElement>('inputRef')
const style = useCssModule()

const classes = computed(() => [
  style.input,
  { [style.invalid]: invalid, [style.success]: success },
])

defineExpose({ inputRef })
</script>

<template>
  <div :class="classes">
    <slot name="prefix" />
    <input ref="inputRef" type="text" v-bind="$attrs" />
    <slot name="suffix" />
  </div>
</template>

<style module>
.input {
  display: flex;
  gap: var(--ga-size-spacing-03);
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-primary);
  border-radius: var(--ga-radius);

  background-color: var(--ga-color-bg-primary);
  padding: var(--ga-size-8) var(--ga-size-12);
  height: var(--ga-size-40);

  &:hover {
    border-color: var(--ga-color-border-action-hover);
  }

  &:has(input:disabled) {
    border-color: var(--ga-color-border-primary);
    background-color: var(--ga-color-surface-disabled);
    color: var(--ga-color-text-disable-selected);

    &:has(input:not(:placeholder-shown)) {
      border-color: var(--ga-color-surface-disable-selected);
      background-color: var(--ga-color-surface-disable-selected);
      color: var(--ga-color-text-on-action);
    }
  }

  &:has(input:focus):not(.invalid, .success) {
    outline: 1px solid var(--ga-color-border-focus);
    border-color: var(--ga-color-border-focus);
  }

  &.invalid {
    outline: 1px solid var(--ga-color-border-error);
    border-color: var(--ga-color-border-error);

    &:has(input:not(:placeholder-shown)) {
      border-color: var(--ga-color-surface-error);
    }
  }

  &.success {
    outline: 1px solid var(--ga-color-border-success);
    border-color: var(--ga-color-border-success);

    &:has(input:not(:placeholder-shown)) {
      background-color: var(--ga-color-surface-success);
    }
  }

  > input {
    outline: none;
    border: none;
    background-color: transparent;
    width: 100%;

    &::placeholder {
      color: var(--ga-color-text-disabled);
    }
  }
}
</style>
