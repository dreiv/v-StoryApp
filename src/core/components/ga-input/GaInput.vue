<script setup lang="ts">
import {
  computed,
  inject,
  useAttrs,
  useCssModule,
  useTemplateRef,
  type InputHTMLAttributes,
} from 'vue'
import { formFieldKey } from '../ga-form-field/types'

defineOptions({ inheritAttrs: false })

export interface GaInputProps extends /* @vue-ignore */ InputHTMLAttributes {
  id?: string
  error?: boolean
  success?: boolean
}
const { id, error, disabled, success } = defineProps<GaInputProps>()
const inputRef = useTemplateRef<HTMLElement>('inputRef')
const style = useCssModule()
const attrs = useAttrs()
const model = defineModel<string>()

const formField = inject(formFieldKey)

const inputAttrs = computed(() => ({
  ...attrs,
  id: id || formField?.id,
  disabled: disabled || formField?.disabled,
}))

const classes = computed(() => [
  style.input,
  { [style.error]: error || formField?.error, [style.success]: success },
])

defineExpose({ inputRef })
</script>

<template>
  <div :class="classes">
    <slot name="prefix" />
    <input ref="inputRef" type="text" v-model="model" v-bind="inputAttrs" />
    <slot name="suffix" />
  </div>
</template>

<style module>
.input {
  display: flex;
  align-items: center;
  gap: var(--ga-size-spacing-03);

  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-primary);
  border-radius: var(--ga-radius);

  background-color: var(--ga-color-bg-primary);
  padding: var(--ga-size-spacing-03) var(--ga-size-spacing-04);
  height: var(--ga-size-spacing-08);

  &:hover {
    border-color: var(--ga-color-border-action-hover);
  }

  &:has(input:disabled) {
    border-color: var(--ga-color-border-primary);
    background-color: var(--ga-color-surface-disabled);
    color: var(--ga-color-text-disable-selected);

    &:has(input:not(:placeholder-shown)) {
      border-color: var(--ga-color-surface-disable-selected);
      background-color: var(--ga-color-surface-disable-selected);
      color: var(--ga-color-text-on-action);
    }
  }

  &:has(input:focus):not(.error, .success) {
    outline: 1px solid var(--ga-color-border-focus);
    border-color: var(--ga-color-border-focus);
  }

  &.error {
    outline: 1px solid var(--ga-color-border-error);
    border-color: var(--ga-color-border-error);

    &:has(input:not(:placeholder-shown)) {
      border-color: var(--ga-color-surface-error);
    }
  }

  &.success {
    outline: 1px solid var(--ga-color-border-success);
    border-color: var(--ga-color-border-success);

    &:has(input:not(:placeholder-shown)) {
      background-color: var(--ga-color-surface-success);
    }
  }

  > input {
    outline: none;
    border: none;
    background-color: transparent;
    width: 100%;

    &::placeholder {
      color: var(--ga-color-text-disabled);
    }
  }
}
</style>
