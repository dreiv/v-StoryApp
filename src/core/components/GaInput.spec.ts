import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaInput from './GaInput.vue'

describe('GaInput', () => {
  let wrapper: VueWrapper

  it('should render correctly', () => {
    wrapper = shallowMount(GaInput)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render prefix slot content', () => {
    wrapper = shallowMount(GaInput, {
      slots: { prefix: '<span data-test="prefix">prefix</span>' },
    })
    expect(wrapper.find('[data-test="prefix"]').exists()).toBe(true)
  })

  it('should render suffix slot content', () => {
    wrapper = shallowMount(GaInput, {
      slots: { suffix: '<span data-test="suffix">suffix</span>' },
    })
    expect(wrapper.find('[data-test="suffix"]').exists()).toBe(true)
  })

  it.each([
    { propName: 'invalid', className: 'invalid' },
    { propName: 'success', className: 'success' },
  ])('should apply $className class when $propName prop is true', ({ propName, className }) => {
    wrapper = shallowMount(GaInput, { props: { [propName]: true } })
    expect(wrapper.find('div').classes()).toContain(className)
  })

  it('should bind value with v-model', async () => {
    wrapper = shallowMount(GaInput, {
      props: { modelValue: 'initial' },
    })

    const input = wrapper.find('input')
    await input.setValue('updated')

    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual(['updated'])
  })
})
