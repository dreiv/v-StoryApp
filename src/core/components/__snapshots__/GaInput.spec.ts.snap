// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaInput > should render correctly 1`] = `"<div class="input"><input type="text"></div>"`;
